src/bigint.rs

```diff
@@ -467,146 +530,323 @@ impl Unsigned for BigUint {}
+        if carry != 0 {
+            self.data.push(carry);
         }
 
-        assert!(borrow == 0, "arithmetic operation overflowed");
-        BigUint::new(diff)
+        self
     }
 }
 
+forward_all_binop_to_val_ref!(impl Sub for BigUint, sub);
+
+fn sub2(a: &mut [BigDigit], b: &[BigDigit]) {
+    let mut b_iter = b.iter();
+    let mut borrow = 0;
+
+    for ai in a.iter_mut() {
+        if let Some(bi) = b_iter.next() {
+            *ai = sbb(*ai, *bi, &mut borrow);
+        } else if borrow != 0 {
+            *ai = sbb(*ai, 0, &mut borrow);
+        } else {
+            break;
+        }
+    }
 
-forward_all_binop_to_val_ref_commutative!(impl Mul for BigUint, mul);
+    /* note: we're _required_ to fail on underflow */
+    assert!(borrow == 0 && b_iter.all(|x| *x == 0),
+            "Cannot subtract b from a because b is larger than a.");
+}
 
-impl<'a> Mul<&'a BigUint> for BigUint {
+impl<'a> Sub<&'a BigUint> for BigUint {
     type Output = BigUint;
 
-    fn mul(self, other: &BigUint) -> BigUint {
-        if self.is_zero() || other.is_zero() { return Zero::zero(); }
-
-        let (s_len, o_len) = (self.data.len(), other.data.len());
-        if s_len == 1 { return mul_digit(other.clone(), self.data[0]);  }
-        if o_len == 1 { return mul_digit(self,  other.data[0]); }
-
-        // Using Karatsuba multiplication
-        // (a1 * base + a0) * (b1 * base + b0)
-        // = a1*b1 * base^2 +
-        //   (a1*b1 + a0*b0 - (a1-b0)*(b1-a0)) * base +
-        //   a0*b0
-        let half_len = cmp::max(s_len, o_len) / 2;
-        let (s_hi, s_lo) = cut_at(self,  half_len);
-        let (o_hi, o_lo) = cut_at(other.clone(), half_len);
-
-        let ll = &s_lo * &o_lo;
-        let hh = &s_hi * &o_hi;
-        let mm = {
-            let (s1, n1) = sub_sign(s_hi, s_lo);
-            let (s2, n2) = sub_sign(o_hi, o_lo);
-            match (s1, s2) {
-                (Equal, _) | (_, Equal) => &hh + &ll,
-                (Less, Greater) | (Greater, Less) => &hh + &ll + (n1 * n2),
-                (Less, Less) | (Greater, Greater) => &hh + &ll - (n1 * n2)
-            }
-        };
+    fn sub(mut self, other: &BigUint) -> BigUint {
+        sub2(&mut self.data[..], &other.data[..]);
+        self.normalize()
+    }
+}
+
+fn sub_sign(a: &[BigDigit], b: &[BigDigit]) -> BigInt {
+    // Normalize:
+    let a = &a[..a.iter().rposition(|&x| x != 0).map_or(0, |i| i + 1)];
+    let b = &b[..b.iter().rposition(|&x| x != 0).map_or(0, |i| i + 1)];
 
-        return ll + mm.shl_unit(half_len) + hh.shl_unit(half_len * 2);
+    match cmp_slice(a, b) {
+        Greater => {
+            let mut ret = BigUint::from_slice(a);
+            sub2(&mut ret.data[..], b);
+            BigInt::from_biguint(Minus, ret.normalize())
+        },
+        Less    => {
+            let mut ret = BigUint::from_slice(b);
+            sub2(&mut ret.data[..], a);
+            BigInt::from_biguint(Minus, ret.normalize())
+        },
```

Comment:
On Thu, Dec 10, 2015 at 04:06:49PM -0800, Josh Stone wrote:
> >  
> > -        return ll + mm.shl_unit(half_len) + hh.shl_unit(half_len * 2);
> > +    match cmp_slice(a, b) {
> > +        Greater => {
> > +            let mut ret = BigUint::from_slice(a);
> > +            sub2(&mut ret.data[..], b);
> > +            BigInt::from_biguint(Minus, ret.normalize())
> > +        },
> > +        Less    => {
> > +            let mut ret = BigUint::from_slice(b);
> > +            sub2(&mut ret.data[..], a);
> > +            BigInt::from_biguint(Minus, ret.normalize())
> > +        },
> 
> Surely one of these branches should be `Plus`?  If so, we have insufficient test coverage to notice this, so please add a new test for it too.  That might be best as a direct `sub_sign` unit test, so it's not relying on particular multiplication details to reach this point.
>
> But you had this right before, and multiplication tests still passed, so I wonder how much that calling code is being exercised.  In fact, I just put a `panic!` in `sub_sign`, and it wasn't reached at all until I enabled the torture test.  With that test we can also see the plain failure here.

*swear* A bug exactly there is why I overloaded add/subtract for slices...
pretty much reintroduced the same bug when I took it back out. Augh.

Yeah, part of the problem is that karatsuba multiplication doesn't become
worthwhile until the numbers are decently sized.

Does rust support gprof yet? code coverage would be really good to have for code
like this.

> If we reduce the iterations of that torture test, we can leave it enabled even in debug testing.  Just 1000 iterations still triggers this particular issue.

Ok yeah, good idea. Maybe I'll keep a second version that's #ignored that runs
lots more iterations.